server/src/main/java/org/apache/druid/server/coordinator/SegmentStateManager.java

```diff
@@ -0,0 +1,261 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.server.coordinator;
+
+import com.google.inject.Inject;
+import org.apache.druid.client.ServerInventoryView;
+import org.apache.druid.java.util.emitter.EmittingLogger;
+import org.apache.druid.metadata.SegmentsMetadataManager;
+import org.apache.druid.timeline.DataSegment;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Manages state of segments being loaded.
+ */
+public class SegmentStateManager
+{
+  private static final EmittingLogger log = new EmittingLogger(SegmentStateManager.class);
+
+  private final LoadQueueTaskMaster taskMaster;
+  private final ServerInventoryView serverInventoryView;
+  private final SegmentsMetadataManager segmentsMetadataManager;
+
+  private final ConcurrentHashMap<String, TierLoadingState> currentlyMovingSegments =
+      new ConcurrentHashMap<>();
+  private final ConcurrentHashMap<String, TierLoadingState> currentlyReplicatingSegments
+      = new ConcurrentHashMap<>();
+
+  @Inject
+  public SegmentStateManager(
+      ServerInventoryView serverInventoryView,
+      SegmentsMetadataManager segmentsMetadataManager,
+      LoadQueueTaskMaster taskMaster
+  )
+  {
+    this.serverInventoryView = serverInventoryView;
+    this.segmentsMetadataManager = segmentsMetadataManager;
+    this.taskMaster = taskMaster;
+  }
+
+  /**
+   * Queues load of a replica of the segment on the given server.
+   */
+  public boolean loadSegment(
+      DataSegment segment,
+      ServerHolder server,
+      boolean isPrimary,
+      ReplicationThrottler throttler
+  )
+  {
+    final String tier = server.getServer().getTier();
+    final LoadPeonCallback callback;
+    if (isPrimary) {
+      // Primary replicas are not subject to throttling
+      callback = null;
+    } else if (canLoadReplica(tier, throttler)) {
+      throttler.incrementAssignedReplicas(tier);
+
+      final TierLoadingState replicatingInTier = currentlyReplicatingSegments
+          .computeIfAbsent(tier, t -> new TierLoadingState(throttler.getMaxLifetime()));
+      replicatingInTier.addSegment(segment.getId(), server.getServer().getHost());
+      callback = success -> replicatingInTier.removeSegment(segment.getId());
+    } else {
+      throttler.incrementThrottledReplicas(tier);
+      return false;
+    }
+
+    try {
+      if (!server.startOperation(segment, SegmentState.LOADING)) {
+        return false;
+      }
+
+      server.getPeon().loadSegment(
+          segment,
+          isPrimary ? SegmentAction.LOAD_AS_PRIMARY : SegmentAction.LOAD_AS_REPLICA,
```

Comment:
Let's please kill this false notion of primary/replica and use words that actually align with the semantics that should exist.

Reply:
Okay, I will include those changes in this PR itself. I am going with the term LOAD vs PRIORITY_LOAD.

Reply:
Fixed.